modules/mod-deep-learning/DeepModelManager.cpp

```diff
@@ -0,0 +1,554 @@
+/**********************************************************************
+
+   Audacity: A Digital Audio Editor
+   Audacity(R) is copyright (c) 1999-2021 Audacity Team.
+
+   DeepModelManager.cpp
+   Hugo Flores Garcia
+
+******************************************************************/
+
+#include "DeepModelManager.h"
+
+#include "CodeConversions.h"
+#include "NetworkManager.h"
+#include "IResponse.h"
+#include "Request.h"
+
+#include <wx/tokenzr.h>
+#include <wx/log.h>
+
+using namespace audacity;
+ 
+DeepModelManager::DeepModelManager() :
+   mAPIEndpoint("https://huggingface.co/api/")
+{
+   const std::string schemaPath = audacity::ToUTF8(
+      wxFileName(DeepModel::BuiltInModulesDir(), wxT("modelcard-schema.json"))
+                  .GetFullPath()
+   );
+   mModelCardSchema = parsers::ParseFile(schemaPath);
+}
+
+DeepModelManager& DeepModelManager::Get()
+{
+   static DeepModelManager manager;
+   return manager;
+}
+
+FilePath DeepModelManager::GetRepoDir(ModelCardHolder card) const
+{
+   static const std::string sep = "_";
+
+   FilePath repoDir = FileNames::MkDir( 
+      wxFileName(DeepModel::DLModelsDir(),
+                 card->author() + sep + card->name() 
+      ).GetFullPath()
+   );
+
+   return repoDir;
+}
+
+DeepModelHolder DeepModelManager::GetModel(ModelCardHolder card) const
+{
+   if (!IsInstalled(card))
+      throw ModelManagerException(XO("Model is not loaded."));
+
+   DeepModelHolder model = std::make_shared<DeepModel>();
+   model->SetCard(card);
+
+   // GetRepoDir won't work if the card is empty
+   wxFileName path = wxFileName(GetRepoDir(card), "model.pt");
+
+   // finally, load
+   model->Load(audacity::ToUTF8(path.GetFullPath()));
+
+   return model;
+}
+
+std::string DeepModelManager::GetMoreInfoURL(ModelCardHolder card) const
+{
+   return "https://huggingface.co/" + card->GetRepoID();
+}
+
+bool DeepModelManager::IsInstalled(ModelCardHolder card) const
+{
+   FilePath repoDir = wxFileName(card->local_path()).GetFullPath();
+
+   wxFileName modelPath = wxFileName(repoDir, "model.pt");
+   wxFileName metadataPath = wxFileName(repoDir, "metadata.json");
```

Comment:
I see the string literals "model.pt" and "metadata.json" repeated in many places.  Better to write each literal only once and give it a variable name, then reuse the name.  This matters less inside one .cpp file, but I don't like it when I see long-distance coincidence of number or string literals in different .cpp files.  Might this file have need to grow and split in future?

Reply:
3a45be1d37e3b81b72ce5da358dab578a0144474